c10/core/SymInt.cpp

```diff
@@ -39,14 +39,18 @@ bool SymInt::has_hint() const {
   return toSymNodeImplUnowned()->has_hint();
 }
 
+// It is important that we do a->METHOD(a->wrap_int(*mb)) instead of
+// a->wrap_int(*mb)->METHOD(a). The latter will not work in the case of
+// SingletonSymNodeImpl because a->wrap_int(*mb) produces a ConstantSymNodeImpl
+// which doesn't implement any operations.
 #define DEFINE_BINARY(API, OP, METHOD, RET)                          \
   RET SymInt::API(const SymInt& sci) const {                         \
     if (auto ma = maybe_as_int()) {                                  \
       if (auto mb = sci.maybe_as_int()) {                            \
         return RET(OP(*ma, *mb));                                    \
       } else {                                                       \
         auto b = sci.toSymNode();                                    \
-        return RET(b->wrap_int(*ma)->METHOD(b));                     \
+        return RET(b->METHOD(b->wrap_int(*ma)));                     \
```

Comment:
This is the reason for most (if not all?) failures in CI right now. (It changes assymetric ops like gt to lt and vice versa, whoops!).

I'm still thinking through different ways to do this, but properly doing multiple dispatch could be one way...

Reply:
I did the simplest thing possible which was to override eq in ConstantSymNode to call into SingletonSymNode. This doesn't do too much in this version, just gives us a slightly nicer error instead of the default "NYI".